File: Cpp/2680_maximum_or.cpp
// https://leetcode.cn/problems/maximum-or/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    long long maximumOr(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> rightor(n, 0);
        for (int i = n-2; i >= 0; --i) {
            rightor[i] = rightor[i+1] | nums[i+1];
        }
        long long res = 0, leftor = 0;
        for (int i = 0; i < n; ++i) {
            long long tmp = leftor | ((long long)nums[i] << k) | rightor[i];
            if (tmp > res) res = tmp;
            leftor |= nums[i];
        }
        return res;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {12, 9};
    int k = 1;
    cout << sol.maximumOr(nums, k) << '\n';
    return 0;
}

File: Python3/2680_maximum_or.py
# https://leetcode.cn/problems/maximum-or/

class Solution:
    def maximumOr(self, nums, k):
        """
        Inputs:
            nums: list[int]
            k: int
        Outputs:
            res: int
        """
        n = len(nums)
        rightor = [0 for _ in range(n)]
        for i in range(n-2, -1, -1):
            rightor[i] = rightor[i+1] | nums[i+1]
        
        res = 0
        leftor = 0
        for i in range(n):
            tmp = leftor | (nums[i] << k) | rightor[i]
            if tmp > res:
                res = tmp
            leftor |= nums[i]
        return res